device/sycl/include/sycl/seeding/detail/sycl_helper.hpp

```diff
@@ -0,0 +1,141 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021-2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <CL/sycl.hpp>
+
+namespace traccc {
+namespace sycl {
+
+// Short aliast for accessor to local memory (shared memory in CUDA)
+template <typename T>
+using local_accessor = ::sycl::accessor<T, 1, ::sycl::access::mode::read_write,
+                                        ::sycl::access::target::local>;
+
+// Some useful helper funcitons
+struct sycl_helper {
+
+    /// Function that performs reduction on the local memory using subgroup
+    /// operations
+    ///
+    /// @param array pointer to an array in local memory
+    /// @param item sycl nd_item for quering local indexes and groups
+    static void reduceInShared(
+        ::sycl::multi_ptr<int, ::sycl::access::address_space::local_space>
+            array,
+        ::sycl::nd_item<1>& item) {
+        auto workItemIdx = item.get_local_id(0);
+        auto sg = item.get_sub_group();
+        auto workGroup = item.get_group();
+        auto groupDim = item.get_local_range(0);
+
+        // Comment out the first two lines of shift_left for Intel platform (min
+        // blockSize is 8 in that case)
+        array[workItemIdx] +=
+            ::sycl::shift_group_left(sg, array[workItemIdx], 16);
+        array[workItemIdx] +=
+            ::sycl::shift_group_left(sg, array[workItemIdx], 8);
+        array[workItemIdx] +=
+            ::sycl::shift_group_left(sg, array[workItemIdx], 4);
+        array[workItemIdx] +=
+            ::sycl::shift_group_left(sg, array[workItemIdx], 2);
+        array[workItemIdx] +=
+            ::sycl::shift_group_left(sg, array[workItemIdx], 1);
+
+        ::sycl::group_barrier(workGroup);
+
+        if (workItemIdx == 0) {
+            for (int i = 1; i < groupDim / 32; i++) {
+                array[workItemIdx] += array[i * 32];
+            }
+        }
+    }
```

Comment:
I will play with this function a little more and try making it work on the Intel devices, but this might have to be replaced by some different reduction techniques. Although, other ones that I've implemented so far don't fully work on different platforms either :confused:

Reply:
Once this version of the code is in, it should become easier for others (like me :wink:) to play with this as well. We'll just do that...